### ISSUE TITLE
Fetcher: PRA

### ISSUE THREAD

**Issue by christinklez** [ucldc/rikolti#266]
FD: https://help.oac.cdlib.org/a/tickets/72909

Mappers dependent on this fetcher:
- PreservicaAPIMapper(Mapper)
***
- Some general links to know:
  - [Development contribution process](https://github.com/ucldc/rikolti#development-contribution-process)
  - ["CIC all tasks" list view](https://github.com/orgs/ucldc/projects/1/views/20)
  - ["CIC current sprint" status view](https://github.com/orgs/ucldc/projects/1/views/7)

- Specific to this fetcher:
  - List of collections, by fetcher type: https://registry.cdlib.org/?harvest_type=PRA
***
- Rikolti MVP project board cheat sheet:
  - Please @ CDL developers liberally, in the comments.
  - When **picking up an "issue,"** please indicate: 
    - a) Assignees: self assign; 
    - b) Sprint: select the current sprint cycle; 
    - c) Date started: indicate the start date -- see right side fields
  - CIC developers will likely need to **create PRs** (and possibly additional issues). When doing so, please also indicate the following: 
    - a) Repository: Rikolti; 
    - b) Milestone: "4 CIC work" (these details are required in order for them to surface in the Rikolti MVP project board); and 
    - c) link the PR to the associated issue.
  - When **requesting review,** please:  
    - a) include a comment & @ CDL developers; and 
    - b) Date delivered: indicate the date delivered (field on the right)

**Comment by christinklez** [comment]
Discussed on 3/27 to pause on the Preservica fetcher/mappers. SO types require multiple API requests. CDL to check-in with OPL and Preservica on both PRA (api) and PRE (atom feed) fetchers to potentially streamline operations. Also, an opportunity to investigate an oai option with Preservica.

**Comment by christinklez** [comment]
@lthurston we have an update on this! Short story is that we will only be using the PRA fetcher (and associated API mapper). Let's plan to discuss during our check-in tomorrow. Thanks!

**Comment by christinklez** [comment]
Note that this set now includes collections that were formally part of the PRE fetcher. There are currently 41 published collections.

**Comment by christinklez** [comment]
List of collections (note that 6 collections have errors popping up with our current harvester): https://docs.google.com/spreadsheets/d/1wiRoArPqW3cAbpPdFFKY6I9drt2Vc0-5u2YMIJVZAu4/edit#gid=143717916

**Comment by christinklez** [comment]
From Preservica:

To answer this, OAI-PMH hasn't changed but we also don't use it so much these days, preferring instead to use the Entity API and Content API. Nevertheless, one of our Senior Customer Success Engineers were able to provide additional information on this. Please see their notes below:

It looks like the two collections are hosted on a common tenant.

The way to differentiate between the two would be to tag the first collection with unique tag (or tags) and then tag the second collection also with unique tag or tags and then create two "service accounts" in Preservica, one with access to the first collection's tag(s), and one with access to the second collections tags(s).

That would allow the first account to only see only OAI-PMH content from the first collection, and so on.

You might also have to ensure that you handle any anonymous access scenarios (in place of a common "public" tag). My recommendation would be to map out any security tag changes and ensure that you have the necessary additional roles in place to support the additional unique collection specific tags.

**Comment by barbarahui** [comment]
@christinklez ~~It sounds like the contributors would need to add a tag to each collection in order for us to use the OAI/PMH endpoint. Is it OK for me to use the Entity and Content APIs as they recommend, or is there a reason we want to use the OAI feed~~

**Comment by barbarahui** [comment]
@christinklez Sorry nevermind! I see that the fetcher that Lucas started writing is already using the entity API.

**Comment by barbarahui** [comment]
I'm getting a 401 unauthorized error when trying to fetch metadata from the preservica API. For example, I'm using this URL for collection [28444](https://registry.cdlib.org/admin/library_collection/collection/28444/change/?_changelist_filters=harvest_type__exact%3DPRA):

https://us.preservica.com/api/entity/v6.0/structural-objects/SO_fc520df3-b3a9-4d82-868a-fca16bb01980/children?max=1

and supplying the basic auth credentials stored in the "Harvest extra data" field in the registry.

Do you know if we have new credentials?

**Comment by christinklez** [comment]
I'll reach out to ask if anything has changed since we last looked at this 3(!) years ago!

**Comment by christinklez** [comment]
@barbarahui - heard back from OPL, regarding the API token:

```
I think you may be able to do this on this page, choosing the "login" option, clicking the arrow to show the dropdown menu, and clicking "try it out": https://demo.preservica.com/api/accesstoken/documentation.html#/

I tried to generate one that way using the username and password you gave me a few days ago but got this error message, which makes me think you may have changed the password when you reset it: 

 "message": "invalid.user Bad credentials",  
"user": "oacops@cdlib.org"

Let me know if this works when you use the current login credentials. 

I'm not at all sure this is the correct method or that it will generate the correct token, so I can poke around more if needed. 
```

OPL has our old password, but obviously we now have a new one! Would you be able to try this out? Or, let me know if this is something I should do!

**Comment by barbarahui** [comment]
I'm not able to enter username, password or tenant into any of the fields. Are you?

**Comment by barbarahui** [comment]
Does the person from OPL know their tenant ID? I think that's what I need.

**Comment by christinklez** [comment]
Sorry, the form fields are not clickable for me. I asked them to generate for us - I needed to send them our new pw, so much for security!

**Comment by barbarahui** [comment]
Oh! So unintuitive!!

**Comment by christinklez** [comment]
Oh actually, you can click "try it out" in order to enable the form fields, whoops. 

I got this error 401 message:
```
{
  "success": false,
  "message": "invalid.user Bad credentials",
  "user": "oacops@cdlib.org"
}
```

I will ask for the tenant

**Comment by barbarahui** [comment]
Yeah this form will only generate a token that's valid for 15 minutes, same as we will be doing in our code via the API, so asking them to generate a code for us wouldn't be helpful anyway except for proving that it's possible. I think the missing piece of info is the tenant ID? I hope so anyway. The documentation is confusing.

**Comment by christinklez** [comment]
Try this URL instead!
https://us.preservica.com/api/accesstoken/documentation.html#//post_login

Thank you!

**Comment by barbarahui** [comment]
~~Heads up that it doesn't look like I can maintain the same Calisphere URLs that we had before. The legacy code, which harvested via the OAI endpoint, mapped the `preservica-id` field as the calisphere id, but I'm not sure that this exists in the metadata we're getting from the content API.~~

~~For example, this record currently in Calisphere:~~

~~https://calisphere.org/item/f1e1333ae8397abc542b8fabda1b5888/~~

~~Is this record in preservica:~~

~~https://oakland.access.preservica.com/uncategorized/IO_27ae225f-505e-40a2-a3f4-6f0e6287e892/~~

~~I don't see `f1e1333ae8397abc542b8fabda1b5888` anywhere on OPL's page, nor in the metadata that's fetched via the API, but I'm only fetching the `id` and the dc fields. Maybe we can ask OPL to ask Preservica if that ID exists anywhere? I looked at the list of available fields and there isn't one called `preservica-id` but maybe they can point us to it.~~

**Comment by christinklez** [comment]
Thanks for the heads up! 

I'm unsure if this is helpful (or at all relevant!), but the legacy harvester supported two methods of fetch/mapping from Preservica. There was the Preservica API (legacy harvester fetcher name: PRA) - which is the method used by the majority of OPL's collections (I believe). There was also the Preservica Atom feed(?) (legacy harvester fetcher name: PRE) - which was used by a handful or fewer OPL collections (I believe).

During Rikolti development, we made the decision that we should only choose one method for Preservica, and opted for PRA (the API method). Here's that decision recorded here: https://github.com/ucldc/rikolti/issues/268

The bummer is that I'm unable to track down which collections used to use PRE.

But here are the legacy mapper codes:
- Preservica API: https://github.com/calisphere-legacy-harvester/dpla-ingestion/blob/ucldc/lib/mappers/preservica_api_mapper.py
- Preservica Atom: https://github.com/calisphere-legacy-harvester/dpla-ingestion/blob/ucldc/lib/mappers/cmis_atom_mapper.py

Anyway, just to get back on topic! I'll reach out to OPL to ask about whether `preservica_id` is an available piece of data they'd be able to expose in the API! Thank you!

**Comment by barbarahui** [comment]
@christinklez D'oh! Sorry, I take that back! I forgot that our legacy fetchers used to do mapping work, so actually the `preservica-id` was based on the `id` that we're still fetching now!! We can actually maintain the same calisphere id after all no problem!
